hedyorg/hedy
# 🪲 Fix error translation and highlighting

Fixes #5240
- highlighing parts of an error is now defined in the error template (in *.po files) using backticks.
- all error templates (in the *.po files) are adjusted to use the new highlighting mechanism.
- duplicated logic for translation and highlighting of errors existed in 3 places in the code base: app.py for errors during transpiling, hedy.py for transpiled errors and statistics.py. The logic is now extracted and unified in hedy_error.py.
- we have tests to check that every error can be formatted with the provided arguments for all languages. However, since we started using safeFormat, these tests could not fail anymore even when there were missing keys. The safeFormat is now removed, the failing tests are corrected and a separate test is added to confirm key errors are raised.

**How to test**
* Automated unit tests are added to cover the new translation and highlighting mechanism, so all tests must pass.
* Choose a subset of errors and check that they appear correctly on the website. For example:
In level 4:
```
is Foobar
print welcome
```
In level 6, supply text value to the ask command instead of a number (note that currently the error should referring to the value 'a' which is not nice but will be addressed in #5005)
```
a = ask 'how much?'
print 360 * a
```
In level 8:
```
repeat 5
```
In level 16:
```
test = 1, 2, 3
```
In level 16. Change the language and ensure that the keyword 'random' in the last example is translated correctly:
```
print list at 12
```
